pars-derive/src/lib.rs

```diff
@@ -222,46 +228,74 @@ fn generate_fmt_block(
-    let _num_fields = cont.data.num_fields();
+    let num_fields = cont.data.num_fields();
+    let field_names = cont.data.to_field_names().unwrap_or_default();
+    let field_names = field_names.iter().map(|s| s.as_str()).collect::<Vec<_>>();
 
     // check that the fmt string is valid
-    let _ = ::pars_fmt::FmtMatcher::new(&fmt_string, field_names.as_slice())
+    let matcher = ::pars_fmt::FmtMatcher::new_named(&fmt_string, field_names.as_slice())
         .map_err(|e| vec![syn::Error::new(Span::call_site(), e)])?;
 
-    let field_names = gen_static_str_slice(field_names.as_slice());
+    let lead_separator = matcher.get_lead_separator_str();
+    let separator_indices = gen_static_array_2tuple(&matcher.make_separator_indices());
 
     let fmt_block = quote! {
-        let field_names = #field_names;
-        let field_names = field_names.to_vec();
+        let lead_separator = #lead_separator;
+        let separator_indices = #separator_indices;
 
-        static INSTANCE: ::pars::OnceCell<::pars::FmtMatcher> = ::pars::OnceCell::INIT;
-        let pat = INSTANCE.get_or_init(|| {
-            ::pars::FmtMatcher::new(#fmt_string, #field_names).unwrap()
-        });
+        let mut pos = 0;
+        let mut current_sep = 0;
+        let mut ordered_matches = [""; #num_fields];
+
+        match src.find(lead_separator) {
+            Some(0) => pos = lead_separator.len(),
+            _ => return Err(::pars::MatchError::missing_separator(current_sep, lead_separator)),
+        };
+
+        for (separator, field_idx) in &separator_indices {
+            if pos == src.len() {
+                return Err(::pars::MatchError::InputExhausted);
+            }
+            if separator.is_empty() {
+                ordered_matches[*field_idx] = &src[pos..];
+                pos = src.len();
+                continue;
+            }
 
-        let ordered_matches = pat.try_match(src)?;
+            match src[pos..].find(separator) {
+                Some(idx) => {
+                    ordered_matches[*field_idx] = &src[pos..pos + idx];
+                    pos = pos + idx + separator.len();
+                }
+                None => return Err(::pars::MatchError::missing_separator(current_sep, separator)),
+            }
+            current_sep += 1;
+        }
     };
     Ok(fmt_block)
 }
 
-/// Creates a static slice of &'static str from a slice of owned Strings.
-fn gen_static_str_slice(inp: &[String]) -> TokenStream {
+/// Generates tokens for a static array from a slice.
+fn gen_static_array<T: quote::ToTokens>(inp: &[T]) -> TokenStream {
     let mut out = TokenStream::new();
-    for _s in inp {
-        out.extend(quote! { #_s, });
+    for item in inp {
+        out.extend(quote! { #item, });
     }
-    quote! {
-        &[#out]
+    quote! { [#out] }
+}
+
+/// Generates tokens for a static array from a slice of 2tuples.
+fn gen_static_array_2tuple<T1, T2>(inp: &[(T1, T2)]) -> TokenStream
```

Comment:
I wonder why `quote` doesn't have blanket implementations for tuples?

Reply:
🤔